src/test/script_tests.cpp

```diff
@@ -1840,8 +1865,79 @@ BOOST_AUTO_TEST_CASE(formatscriptflags)
+    BOOST_CHECK_EQUAL(FormatScriptFlags(SCRIPT_VERIFY_TAPROOT | (1u<<28)), "TAPROOT,0x10000000");
+    BOOST_CHECK_EQUAL(FormatScriptFlags(1u<<28), "0x10000000");
+}
+
+
+void DoTapscriptTest(std::vector<unsigned char> witVerifyScript, std::vector<std::vector<unsigned char>> witData, const std::string& message, int scriptError)
+{
+    const KeyData keys;
+    TaprootBuilder builder;
+    builder.Add(/*depth=*/0, witVerifyScript, TAPROOT_LEAF_TAPSCRIPT, /*track=*/true);
+    builder.Finalize(XOnlyPubKey(keys.key0.GetPubKey()));
+
+    CScriptWitness witness;
+    witness.stack.insert(witness.stack.begin(), witData.begin(), witData.end());
+    witness.stack.push_back(witVerifyScript);
+    auto controlblock = *(builder.GetSpendData().scripts[{witVerifyScript, TAPROOT_LEAF_TAPSCRIPT}].begin());
+    witness.stack.push_back(controlblock);
+
+    uint32_t flags = SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS | SCRIPT_VERIFY_TAPROOT | SCRIPT_VERIFY_OP_CAT;
+    CScript scriptPubKey = CScript() << OP_1 << ToByteVector(builder.GetOutput());
+    CScript scriptSig = CScript(); // Script sig is always size 0 and empty in tapscript
+    DoTest(scriptPubKey, scriptSig, witness, flags, message, scriptError, /*nValue=*/1);
+}
+
+BOOST_AUTO_TEST_CASE(cat_simple)
+{
+    std::vector<std::vector<unsigned char>> witData;
+    witData.push_back(ParseHex("aa"));
+    witData.push_back(ParseHex("bbbb"));
+    std::vector<unsigned char> witVerifyScript = {OP_CAT, OP_PUSHDATA1, 0x03, 0xaa, 0xbb, 0xbb, OP_EQUAL};
+    DoTapscriptTest(witVerifyScript, witData, "Simple CAT", SCRIPT_ERR_OK);
+}
+
+
+BOOST_AUTO_TEST_CASE(cat_empty_stack)
+{
+    // Ensures that OP_CAT successfully handles concatenating two empty stack elements
+    std::vector<std::vector<unsigned char>> witData;
+    witData.push_back({});
+    witData.push_back({});
+    std::vector<unsigned char> witVerifyScript = {OP_CAT, OP_PUSHDATA1, 0x00,OP_EQUAL};
+    DoTapscriptTest(witVerifyScript, witData, "CAT empty stack", SCRIPT_ERR_OK);
+}
+
+BOOST_AUTO_TEST_CASE(cat_dup_test)
+{
+    // CAT DUP exhaustion attacks using all element sizes from 1 to 522
+    // with CAT DUP repetitions up to 10. Ensures the correct error is thrown
+    // or not thrown, as appropriate.
+    unsigned int maxElementSize = 522;
+    unsigned int maxDupsToCheck = 10;
+
+    std::vector<std::vector<unsigned char>> witData;
+    witData.push_back({});
+    for (unsigned int elementSize = 1; elementSize <= maxElementSize; elementSize++) {
+        std::vector<unsigned char> witVerifyScript;
+        // increase the size of stack element by one byte
+        witData.at(0).push_back(0x1A);
+        for (unsigned int dups = 1; dups <= maxDupsToCheck; dups++) {
+            witVerifyScript.push_back(OP_DUP);
+            witVerifyScript.push_back(OP_CAT);
+            int expectedErr = SCRIPT_ERR_OK;
+            unsigned int catedStackElementSize = witData.at(0).size()*pow(2, dups);
```

Comment:
I'm getting a `error: use of undeclared identifier 'pow'` error here; `pow` returns a double not an integer; shouldn't this just be `witData.at(0).size() << dups` ?

Reply:
I dont see why `<< dups` wouldn't work.

Reply:
`./src/test/test_bitcoin --log_level=test_suite --run_test=script_tests`
Passing locally